File: backend_ts/src/services/email.ts
/**
 * 邮件服务 - 从数据库读取 SMTP 配置并发送邮件
 * 对标 Rust backend 的 EmailService
 */

import type { Transporter } from "nodemailer";
import nodemailer from "nodemailer";
import { getDb } from "@/lib/db";

interface EmailConfig {
  host: string;
  port: number;
  user: string;
  password: string;
  fromEmail: string;
  fromName: string;
  encryption: "tls" | "starttls" | "none";
}

/**
 * 从数据库获取邮件配置
 */
async function getEmailConfig(): Promise<EmailConfig> {
  const db = await getDb();
  const collection = db.collection("options");

  const option = await collection.findOne({ name: "email" });
  if (!option || !option.value) {
    throw new Error("邮件配置未找到");
  }

  const config = option.value;

  return {
    host: config.host || "smtp.gmail.com",
    port: config.port || 587,
    user: config.user || "",
    password: config.password || "",
    fromEmail: config.fromEmail || config.user || "",
    fromName: config.fromName || "Neo Space",
    encryption: config.encryption || "starttls",
  };
}

/**
 * 创建邮件传输器
 */
async function createTransporter(): Promise<Transporter> {
  const config = await getEmailConfig();

  const transportOptions: any = {
    host: config.host,
    port: config.port,
    auth: {
      user: config.user,
      pass: config.password,
    },
  };

  // 根据加密方式配置
  if (config.encryption === "tls") {
    transportOptions.secure = true; // 端口 465
  } else if (config.encryption === "starttls") {
    transportOptions.secure = false; // 端口 587
    transportOptions.requireTLS = true;
  } else {
    transportOptions.secure = false;
  }

  return nodemailer.createTransporter(transportOptions);
}

/**
 * 发送验证码邮件
 */
export async function sendVerificationEmail(
  toEmail: string,
  code: string,
  siteName = "Neo Space",
): Promise<void> {
  try {
    const config = await getEmailConfig();
    const transporter = await createTransporter();

    const currentYear = new Date().getFullYear();

    // HTML 邮件内容
    const htmlBody = `
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <style>
        body { margin: 0; padding: 0; font-family: -apple-system, sans-serif; background: #f4f7f6; }
        .container { max-width: 600px; margin: 40px auto; background: #ffffff; border-radius: 16px; overflow: hidden; box-shadow: 0 4px 12px rgba(0, 0, 0, 0.1); }
        .header { background: linear-gradient(135deg, #0d9488 0%, #14b8a6 100%); padding: 30px; text-align: center; color: white; }
        .content { padding: 40px 30px; }
        .code-container { background: #f0fdfa; border: 2px dashed #14b8a6; border-radius: 12px; padding: 20px; text-align: center; margin: 20px 0; }
        .code { font-size: 32px; font-weight: bold; color: #0d9488; letter-spacing: 5px; }
        .footer { background: #f9fafb; padding: 20px; text-align: center; font-size: 12px; color: #6b7280; border-top: 1px solid #e5e7eb; }
    </style>
</head>
<body>
    <div class="container">
        <div class="header"><h1>验证码</h1></div>
        <div class="content">
            <p>你好！</p>
            <p>感谢你访问 <strong>${siteName}</strong>。请使用下方的验证码完成操作：</p>
            <div class="code-container">
                <div class="code">${code}</div>
                <p style="font-size: 12px; color: #666;">有效期 10 分钟</p>
            </div>
            <p style="color: #92400e; font-size: 13px;">⚠️ 如果非本人操作，请忽略此邮件。</p>
        </div>
        <div class="footer">
            <p>© ${currentYear} ${siteName} · Powered by Neo Space</p>
        </div>
    </div>
</body>
</html>
`;

    const textBody = `你好！\n\n你的验证码是：${code}\n\n有效期 10 分钟。\n\n© ${currentYear} ${siteName}`;

    await transporter.sendMail({
      from: `"${config.fromName}" <${config.fromEmail}>`,
      to: toEmail,
      subject: `[${siteName}] 验证码`,
      text: textBody,
      html: htmlBody,
    });

    console.log(`[Email] 邮件已成功发送至: ${toEmail}`);
  } catch (error) {
    console.error(`[Email] 邮件发送失败:`, error);
    throw new Error(`邮件发送失败: ${error}`);
  }
}

/**
 * 发送友链申请通知邮件
 */
export async function sendLinkApplicationNotification(
  adminEmail: string,
  linkName: string,
  linkUrl: string,
): Promise<void> {
  try {
    const config = await getEmailConfig();
    const transporter = await createTransporter();

    const htmlBody = `
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <style>
        body { font-family: -apple-system, sans-serif; line-height: 1.6; }
        .container { max-width: 600px; margin: 0 auto; padding: 20px; }
        .header { background: #0d9488; color: white; padding: 20px; border-radius: 8px 8px 0 0; }
        .content { background: #f9fafb; padding: 20px; border-radius: 0 0 8px 8px; }
        .link-info { background: white; padding: 15px; border-radius: 8px; margin: 15px 0; }
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h2>新的友链申请</h2>
        </div>
        <div class="content">
            <p>你好！</p>
            <p>有新的友链申请需要审核：</p>
            <div class="link-info">
                <p><strong>友链名称：</strong>${linkName}</p>
                <p><strong>友链地址：</strong><a href="${linkUrl}">${linkUrl}</a></p>
            </div>
            <p>请登录后台进行审核。</p>
        </div>
    </div>
</body>
</html>
`;

    await transporter.sendMail({
      from: `"${config.fromName}" <${config.fromEmail}>`,
      to: adminEmail,
      subject: `[友链申请] ${linkName}`,
      html: htmlBody,
    });

    console.log(`[Email] 友链申请通知已发送至: ${adminEmail}`);
  } catch (error) {
    console.error(`[Email] 友链申请通知发送失败:`, error);
    throw new Error(`邮件发送失败: ${error}`);
  }
}

File: backend/src/config/email.rs
//! Email configuration from database

use mongodb::{bson::doc, Database};
use std::error::Error;

/// SMTP 加密方式
#[derive(Debug, Clone, PartialEq)]
pub enum SmtpEncryption {
    /// STARTTLS (端口 587)
    StartTls,
    /// 直接 TLS/SSL (端口 465)
    Tls,
    /// 无加密 (不推荐)
    None,
}

/// 邮箱配置（从数据库读取）
#[derive(Debug, Clone)]
pub struct EmailConfig {
    pub host: String,
    pub port: u16,
    pub user: String,
    pub password: String,
    pub from_name: String,
    pub from_email: String,
    pub encryption: SmtpEncryption,
}

/// 从数据库读取邮箱配置
pub async fn get_email_config(db: &Database) -> Result<EmailConfig, Box<dyn Error>> {
    let collection: mongodb::Collection<mongodb::bson::Document> = db.collection("options");

    let option = collection
        .find_one(doc! { "name": "mailOptions" })
        .await?
        .ok_or("邮箱配置不存在")?;

    let value = option
        .get_document("value")
        .map_err(|_| "邮箱配置格式错误")?;

    // 检查是否启用邮件功能
    if !value.get_bool("enable").unwrap_or(false) {
        return Err("邮件功能未启用".into());
    }

    // 读取 options 子文档
    let options = value
        .get_document("options")
        .map_err(|_| "邮箱 options 配置缺失")?;

    let user = value
        .get_str("user")
        .map_err(|_| "邮箱用户名未配置")?
        .to_string();
    let from_email = value.get_str("from").unwrap_or(&user).to_string();

    let port = u16::try_from(options.get_i32("port").unwrap_or(587))
        .map_err(|_| "邮箱端口号无效")?;

    let secure = options.get_bool("secure").unwrap_or(false);

    // 修正后的加密逻辑判断
    let encryption = if port == 465 {
        // 465 端口几乎永远是隐式 TLS
        SmtpEncryption::Tls
    } else if port == 587 {
        // 587 端口几乎永远是 STARTTLS
        SmtpEncryption::StartTls
    } else if secure {
        // 如果是其他非标准端口但标记了 secure，尝试隐式 TLS
        SmtpEncryption::Tls
    } else if port == 25 {
        SmtpEncryption::None
    } else {
        // 默认 fallback 到 STARTTLS
        SmtpEncryption::StartTls
    };

    Ok(EmailConfig {
        host: options
            .get_str("host")
            .map_err(|_| "SMTP 主机未配置")?
            .to_string(),
        port,
        user: user.clone(),
        password: value
            .get_str("pass")
            .map_err(|_| "邮箱密码未配置")?
            .to_string(),
        from_name: "Neo Space".to_string(),
        from_email,
        encryption,
    })
}
